gpflow/models/training_mixins.py

```diff
@@ -0,0 +1,129 @@
+        various optimizers, namely :class:`gpflow.optimizers.Scipy` and subclasses of
+        `tf.optimizers.Optimizer`.
+
+    See :class:`ExternalDataTrainingLossMixin` for an equivalent mixin for models that do **not**
+    own their own data.
+    """
+
+    def training_loss(self) -> tf.Tensor:
+        """
+        Returns the training loss for this model.
+        """
+        return self._training_loss()
+
+    def training_loss_closure(self, jit=True) -> Callable[[], tf.Tensor]:
+        """
+        Convenience method. Returns a closure which itself returns the training loss. This closure
+        can be passed to the minimize methods on :class:`gpflow.optimizers.Scipy` and subclasses of
+        `tf.optimizers.Optimizer`.
+
+        :param jit: If `True` (default), compile the training loss function in a TensorFlow graph
+            by wrapping it in tf.function()
+        """
+        if jit:
+            return tf.function(self.training_loss)
+        return self.training_loss
+
+
+class ExternalDataTrainingLossMixin:
+    """
+    Mixin utility for training loss methods for models that do **not** own their own data.
+    It provides
+
+      - a uniform API for the training loss :meth:`training_loss`
+      - a convenience method :meth:`training_loss_closure` for constructing the closure expected by
+        various optimizers, namely :class:`gpflow.optimizers.Scipy` and subclasses of
+        `tf.optimizers.Optimizer`.
+
+    See :class:`InternalDataTrainingLossMixin` for an equivalent mixin for models that **do** own
+    their own data.
+    """
+
+    def training_loss(self, data: Data) -> tf.Tensor:
+        """
+        Returns the training loss for this model.
+        
+        :param data: the data to be used for computing the model objective.
+        """
+        return self._training_loss(data)
+
+    def training_loss_closure(
+        self, data: Union[Data, Iterator[Data]], jit=True, input_signature=None,
+    ) -> Callable[[], tf.Tensor]:
+        """
+        Returns a closure that computes the training loss, which by default is
+        wrapped in tf.function(). This can be disabled by passing `jit=False`.
+        
+        :param data: the data to be used by the closure for computing the model
+            objective. Can be the full dataset or an iterator (e.g.
+            `iter(dataset.batch(batch_size))`, where dataset is an instance of
+            tf.data.Dataset)
+        :param jit: whether to wrap training loss in tf.function()
+        :param input_signature: to be passed to tf.function() when jit=True;
+            will be inferred from the iterator if `data` is an iterator on a
+            tf.data.Dataset.
+        """
+        training_loss = self.training_loss
+        if jit:
+            if isinstance(data, DatasetOwnedIterator):
+                tensor_specs = [
+                    tf.TensorSpec(shape, dtype)
+                    for shape, dtype in zip(data.output_shapes, data.output_types)
```

Comment:
these attributes appear to be deprecated ... see https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/data/ops/iterator_ops.py

`tf.compat.v1.data.get_output_shapes`
`tf.compat.v1.data.get_output_types`

Reply:
what do you suggest we replace it with?

Reply:
ah, there's now `element_specs`! much neater

Reply:
thank you, great catch